### PR TITLE
feat: NoteToSelf messages are editable forever

### LINKED ISSUE #13033: Keep messages in "Note to self" chat editable forever
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

## Feature request

**Which Nextcloud Version are you currently using:** NC30

**Is your feature request related to a problem? Please describe.**
Currently, messages are only editable for 24 hours. When using the Note to Self chat as a flexible To Do list (for example), this means I can only ever mark a To Do item as done within my 24 hour window, but not after.

**Describe the solution you'd like**
Since there is no reason why messages that are only ever seen and read by myself should be kept immutable for posterity as they are in chats, make the messages in the "Note to self" editable forever.

**Describe alternatives you've considered**
None - using another app doesn't make me happy :crab:

### PR DESCRIPTION
**Pull request by skalidindi53**
### ☑️ Resolves

* Fix #13033 

### 🚧 Tasks

- [x] NoteToSelf messages are now editable forever

### 🏁 Checklist

- [x] ⛑️ Tests (unit and/or integration) are included or not possible
- [x] 📘 API documentation in `docs/` has been updated or is not required
- [x] 🔖 Capability is added or not needed

### CODE REVIEW

File: lib/Controller/ChatController.php [hunk 1]
```diff
@@ -845,13 +845,15 @@ public function editMessage(int $messageId, string $message): DataResponse {
 			return new DataResponse([], Http::STATUS_METHOD_NOT_ALLOWED);
 		}
 
-		$maxAge = $this->timeFactory->getDateTime();
-		$maxAge->sub(new \DateInterval('P1D'));
-		if ($comment->getCreationDateTime() < $maxAge) {
-			// Message is too old
-			return new DataResponse(['error' => 'age'], Http::STATUS_BAD_REQUEST);
+		if (!($this->room->getType() === Room::TYPE_NOTE_TO_SELF)) {
```
Comment (nickvergessen) [inline lib/Controller/ChatController.php:848]:
```suggestion
		if ($this->room->getType() !== Room::TYPE_NOTE_TO_SELF) {
```

File: tests/integration/features/chat-1/note-to-self.feature [hunk 2]
```diff
@@ -22,3 +22,18 @@ Feature: chat/note-to-self
...
+    Then user "participant1" sends message "Initial message" to room "participant1-note-to-self" with 201
+    And user "participant1" sees the following messages in room "participant1-note-to-self" with 200
+      | room                      | actorType | actorId      | actorDisplayName         | message          | messageParameters |
+      | participant1-note-to-self | users     | participant1 | participant1-displayname | Initial message  | []                |
+    When aging messages 24 hours in room "participant1-note-to-self"
+    And user "participant1" edits message "Initial message" in room "participant1-note-to-self" to "Edited after 24 hours" with 200
+    Then user "participant1" sees the following messages in room "participant1-note-to-self" with 200
+      | room                      | actorType | actorId      | actorDisplayName         | message                | messageParameters | lastEditActorType | lastEditActorId | lastEditActorDisplayName |
+      | participant1-note-to-self | users     | participant1 | participant1-displayname | Edited after 24 hours  | []                | users             | participant1    | participant1-displayname |
```
Comment (nickvergessen) [inline tests/integration/features/chat-1/note-to-self.feature:39]: New line at the end please :)

**Comment by nickvergessen** [comment]
Hmm I guess that means we need frontend and client changes as well?

**Comment by SystemKeeper** [comment]
> Hmm I guess that means we need frontend and client changes as well?

And a capability …

File: lib/Controller/ChatController.php [hunk 3]
```diff
@@ -845,13 +845,15 @@ public function editMessage(int $messageId, string $message): DataResponse {
...
+		if ($this->room->getType() !== Room::TYPE_NOTE_TO_SELF) {
+			$maxAge = $this->timeFactory->getDateTime();
+			$maxAge->sub(new \DateInterval('P1D'));
+			if ($comment->getCreationDateTime() < $maxAge) {
+				// Message is too old
+				return new DataResponse(['error' => 'age'], Http::STATUS_BAD_REQUEST);
+			}
 		}
-
+
```
Comment (nickvergessen) [inline lib/Controller/ChatController.php:856]:
```suggestion

```

File: docs/capabilities.md [hunk 4]
```diff
@@ -154,3 +154,6 @@
 * `ban-v1` - Whether the API to ban attendees is available
 * `mention-permissions` - Whether non-moderators are allowed to mention `@all`
 * `federation-v2` - Whether federated session ids are used and calls are possible with federation
+
+## 21
```
Comment (nickvergessen) [inline docs/capabilities.md:158]:
```suggestion
```

We'll do that for 20

**Review by nickvergessen — APPROVED** [review]

**Comment by SystemKeeper** [comment]
Adjustment of capability test is missing

**Pull request merged**